crates/ra_assists/src/handlers/auto_import.rs

```diff
@@ -81,11 +62,269 @@ pub(crate) fn auto_import(ctx: AssistCtx) -> Option<Assist> {
+                        ModuleDef::Trait(trait_candidate)
+                            if trait_candidate
+                                .items(db)
+                                .into_iter()
+                                .any(|item| item == AssocItem::Function(called_function)) =>
+                        {
+                            Some(trait_candidate)
+                        }
+                        _ => None,
+                    })
+                    .collect::<FxHashSet<_>>()
+            })
+            .flatten()
+            .collect()
+    }
+}
+
+#[derive(Debug)]
+enum ImportCandidate {
+    /// Simple name like 'HashMap'
+    UnqualifiedName(String),
+    /// First part of the qualified name.
+    /// For 'std::collections::HashMap', that will be 'std'.
+    QualifierStart(String),
+    /// A trait associated function (with no self parameter) or associated constant.
+    /// For 'test_mod::TestEnum::test_function', `Type` is the `test_mod::TestEnum` expression type
+    /// and `String` is the `test_function`
+    TraitAssocItem(Type, String),
+    /// A trait method with self parameter.
+    /// For 'test_enum.test_method()', `Type` is the `test_enum` expression type
+    /// and `String` is the `test_method`
+    TraitMethod(Type, String),
+}
+
+impl ImportCandidate {
+    fn for_method_call(
+        method_call: &ast::MethodCallExpr,
+        source_analyzer: &SourceAnalyzer,
+        db: &impl HirDatabase,
+    ) -> Option<Self> {
+        if source_analyzer.resolve_method_call(method_call).is_some() {
+            return None;
+        }
+        Some(Self::TraitMethod(
+            source_analyzer.type_of(db, &method_call.expr()?)?,
+            method_call.name_ref()?.syntax().to_string(),
+        ))
+    }
+
+    fn for_regular_path(
+        path_under_caret: &ast::Path,
+        source_analyzer: &SourceAnalyzer,
+        db: &impl HirDatabase,
+    ) -> Option<Self> {
+        if source_analyzer.resolve_path(db, path_under_caret).is_some() {
+            return None;
+        }
+
+        let segment = path_under_caret.segment()?;
+        if let Some(qualifier) = path_under_caret.qualifier() {
+            let qualifier_start = qualifier.syntax().descendants().find_map(ast::NameRef::cast)?;
+            let qualifier_start_path =
+                qualifier_start.syntax().ancestors().find_map(ast::Path::cast)?;
+            if let Some(qualifier_start_resolution) =
+                source_analyzer.resolve_path(db, &qualifier_start_path)
+            {
+                let qualifier_resolution = if qualifier_start_path == qualifier {
+                    qualifier_start_resolution
+                } else {
+                    source_analyzer.resolve_path(db, &qualifier)?
+                };
+                if let PathResolution::Def(ModuleDef::Adt(function_callee)) = qualifier_resolution {
+                    Some(ImportCandidate::TraitAssocItem(
+                        function_callee.ty(db),
```

Comment:
this won't handle generics correctly; we'll probably need to record the correct type during inference (later)

Reply:
I'm afraid I don't know how to record and use that and need some code pointers.

Reply:
I'm not suggesting we do that in this PR, but I can give some pointers later of course :slightly_smiling_face:

Reply:
Splendid, thanks :)

Reply:
By the way, feel free to create an issue or a Zulip thread with those pointers ;)